packages/catalog-server/src/lib/firestore/firestore-repository.ts

```diff
@@ -0,0 +1,323 @@
+  CustomElement,
+  PackageInfo,
+  PackageStatus,
+  PackageVersion,
+  VersionStatus,
+  ValidationProblem
+} from '@webcomponents/catalog-api/lib/schema.js';
+import {Package} from '@webcomponents/custom-elements-manifest-tools/lib/npm.js';
+import {Repository} from '../repository.js';
+import {packageInfoConverter} from './package-info-converter.js';
+import {packageVersionConverter} from './package-version-converter.js';
+import {customElementConverter} from './custom-element-converter.js';
+import {validationProblemConverter} from './validation-problem-converter.js';
+
+const projectId = 'wc-catalog';
+firebase.initializeApp({projectId});
+export const db = new Firestore({projectId});
+
+export class FirestoreRepository implements Repository {
+  async startPackageVersionImport(
+    packageName: string,
+    version: string
+  ): Promise<void> {
+    // TODO: verify that the package exists. Or does Firestore already do that
+    // with subcollections?
+    const versionRef = getPackageVersionRef(packageName, version);
+    // Since create() fails if the document exists, we don't need a transaction
+    await versionRef.create({
+      status: VersionStatus.INITIALIZING,
+      lastUpdate: FieldValue.serverTimestamp(),
+    });
+  }
+
+  async endPackageVersionImportWithReady(
+    packageName: string,
+    version: string,
+    packageMetadata: Package,
+    customElementsManifestSource: string | undefined
+  ): Promise<void> {
+    const packageVersionMetadata = packageMetadata.versions[version]!;
+    await db.runTransaction(async (t) => {
+      const versionRef = getPackageVersionRef(packageName, version);
+      const packageVersionDoc = await t.get(
+        versionRef.withConverter(packageVersionConverter)
+      );
+      const packageVersionData = packageVersionDoc.data();
+      if (packageVersionData === undefined) {
+        console.error(`Package version not found: ${packageName}@${version}`);
+        throw new Error(`Package version not found: ${packageName}@${version}`);
+      }
+      if (packageVersionData.status !== VersionStatus.INITIALIZING) {
+        console.error(
+          `Unexpected package version status: ${packageVersionData.status}`
+        );
+        throw new Error(
+          `Unexpected package version status: ${packageVersionData.status}`
+        );
+      }
+      const packageTime = packageMetadata.time[version]!;
+      const packageType = packageVersionMetadata.type ?? 'commonjs';
+      const author = packageVersionMetadata.author?.name ?? '';
+      const distTags = packageMetadata['dist-tags'];
+      const versionDistTags = getDistTagsForVersion(distTags, version);
+
+      // Store package data and mark version as ready
+      // TODO: we want this type to match the schema and converter type. How do
+      // we get strongly typed refs?
+      await t.set(versionRef, {
+        status: VersionStatus.READY,
+        lastUpdate: FieldValue.serverTimestamp(),
+        // package: packageName,
```

Comment:
Is this supposed to be commented? package seems important

Reply:
It's not in the GraphQL schema type for PackageVersion, since PackageVersion is always nested under a PackageInfo in the GraphQL schema and is a subcollection of package in the Firestore layout. So the package name is always available there.

What's it's needed for is denormalization for queries, which aren't implemented yet. I can remove or add a comment that this needs to be added and a type created with denormalized fields added to the schema type.

Reply:
Added comment